### PR TITLE
support recursive cte

### PR DESCRIPTION
**Pull request by iamlinjunhong**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [ ] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #9780 

## What this PR does / why we need it:

### CODE REVIEW

File: pkg/container/batch/batch.go
Comment (XuPeng-SH) [inline pkg/container/batch/batch.go:1]: Batch is a basic data structure, which will affect the compatibility of data. If it is modified, it needs to be adapted for compatibility.

**Review by XuPeng-SH — CHANGES_REQUESTED** [review]
changes are needed for compatibility

**Review by zhangxu19830126 — CHANGES_REQUESTED** [review]
All changes after version 0.8 need to be kept compatible. E.g. iota and pb definitions

**Review by heni02 — APPROVED** [review]

**Comment by aunjgr** [comment]
> All changes after version 0.8 need to be kept compatible. E.g. iota and pb definitions

We can't freeze pb definitions for the time being. Pb of both plan and pipeline is changing.

**Review by zhangxu19830126 — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]
I'm OK with this PR. Some modifications on SINK/SINK_SCAN will be done by @zengyan1

**Review by m-schen — APPROVED** [review]
colexec and function directory lgtm.

File: pkg/sql/compile/types.go [hunk 1]
```diff
@@ -199,14 +199,15 @@ type Compile struct {
 
 	s3CounterSet perfcounter.CounterSet
 
-	nodeRegs map[int32]*process.WaitRegister
-	stepRegs map[int32][]int32
+	nodeRegs map[[2]int32]*process.WaitRegister
+	stepRegs map[int32][][2]int32
```
Comment (nnsgmsone) [inline pkg/sql/compile/types.go:203]: 为什么需要一个nodeid参数呢。。

**Review by aunjgr — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]

File: pkg/sql/compile/types.go (hunk 1 above)
Comment (iamlinjunhong) [reply pkg/sql/compile/types.go:203]: 因为递归的 sink scan 需要接收 initial pipeline 的数据和自己的数据，如果没有 nodeid ，initial pipeline 的sink 去 getStepRegs 时会拿到三个发送的地方。正确的应该是两个

**Pull request merged**